bidi.py

```diff
@@ -0,0 +1,339 @@
+
+import asyncio
+import json
+import logging
+import socket
+import tempfile
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any
+
+import aiohttp
+
+log_proto = logging.getLogger("bidi.proto")
+log_command = logging.getLogger("bidi.command")
+
+
+class WebdriverError(RuntimeError):
+    pass
+
+
+class Error(RuntimeError):
+    def __init__(self, msg: str) -> None:
+        self.msg = msg
+
+    def __str__(self) -> str:
+        return self.msg
+
+
+# default timeout
+TIMEOUT = 5  # TODO: raise to 15
+
+
+@dataclass
+class LogMessage:
+    level: str  # like "info"
+    type: str  # usually "console"
+    timestamp: int
+    args: list[object]
+    text: str
+
+    def __init__(self, message_params):
+        self.level = message_params["level"]
+        self.type = message_params["type"]
+        self.timestamp = message_params["timestamp"]
+        self.args = message_params.get("args", [])
+        self.text = message_params["text"]
+
+    def __str__(self):
+        return f"LogMessage: {self.type} {self.level} @{self.timestamp}: {self.text} {self.args}"
+
+
+@dataclass
+class BidiSession:
+    ws_url: str
+    session_url: str
+    process: asyncio.subprocess.Process
+
+
+# Return port numbers that were free at the time of checking
+# They might be in use again by the time the function returns...
+def pick_ports(count: int) -> list[int]:
+    sockets: list[socket.socket] = []
+    ports: list[int] = []
+
+    for _ in range(count):
+        sock = socket.socket()
+        sock.bind(('127.0.0.1', 0))
+        sockets.append(sock)
+        ports.append(sock.getsockname()[1])
+
+    for s in sockets:
+        s.close()
+
+    return ports
+
+
+def jsquote(js: object) -> str:
+    return json.dumps(js)
+
+
+class WebdriverBidi:
+    def __init__(self, headless=False) -> None:
+        self.headless = headless
+        self.last_id = 0
+        self.pending_commands: dict[int, asyncio.Future] = {}
+        self.logs: list[LogMessage] = []
+        self.bidi_session: BidiSession | None = None
+        self.future_wait_page_load = None
+        self.top_context: str | None = None  # top-level browsingContext
+        self.context: str | None  # currently selected context (top or iframe)
+
+    async def start_bidi_session(self) -> None:
+        raise NotImplementedError('must be implemented by concrete subclass')
+
+    async def close_bidi_session(self) -> None:
+        raise NotImplementedError('must be implemented by concrete subclass')
+
+    async def close(self):
+        assert self.bidi_session is not None
+        log_proto.debug("cleaning up webdriver")
+
+        self.task_reader.cancel()
+        del self.task_reader
+        await self.ws.close()
+        await self.close_bidi_session()
+        self.bidi_session.process.terminate()
+        await self.bidi_session.process.wait()
+        self.bidi_session = None
+        await self.http_session.close()
+
+    def ws_done_callback(self, future):
```

Comment:
It probably also makes sense to reject all outstanding commands here.  If something crashed, you don't want to get stuck awaiting a response that will never come.

Reply:
Help, I'm seeing double (bad github day?)